### ISSUE TITLE
【fog-azure-rm】undefined method `copy_object' for Fog::Storage::AzureRM::Real

### ISSUE THREAD

**Issue by sakahiro** [carrierwaveuploader/carrierwave#2410]
environment
rails...5.2.3
carrierwave...2.0.0
fog-azure-rm...0.3.2

I tried to use fog-azure-rm as a provider.
But when I upload images, `undefined method copy_object` error occurred.

My settings in `config/initializers/carrierwave.rb`are like below
```.ruby
CarrierWave.configure do |config|
  config.asset_host = "xxx"

  config.cache_dir = Rails.root.join("tmp", "uploads")

  config.fog_credentials = {
    provider: "AzureRM",
    azure_storage_account_name: "xxx",
    azure_storage_access_key: "xxx",
    }

  config.fog_directory  = true
  config.fog_public     = false
  config.fog_attributes = { "Cache-Control" => "max-age=#{365.day.to_i}" }
end
```

and in my uploader

```.ruby
# frozen_string_literal: true

require 'fog/azurerm'

class BaseUploader < CarrierWave::Uploader::Base
  storage :fog

  ...
end
```

Maybe this method call from [here](https://github.com/carrierwaveuploader/carrierwave/blob/master/lib/carrierwave/storage/fog.rb#L445) but, there is no method in [fog-azure-rm](https://github.com/fog/fog-azure-rm/blob/master/lib/fog/azurerm/storage.rb).

Are my settings wrong?
If not, how do you think to call `copy_blob` from Carrierwave when the provider is AzureRM?

**Comment by stanhu** [comment]
Yeah, I ran into this. I solved this by patching `fog-azure-rm` like so:

```diff
diff --git a/lib/fog/azurerm/requests/storage/copy_blob.rb b/lib/fog/azurerm/requests/storage/copy_blob.rb
index 2c277fe..ac51dae 100644
--- a/lib/fog/azurerm/requests/storage/copy_blob.rb
+++ b/lib/fog/azurerm/requests/storage/copy_blob.rb
@@ -17,6 +17,10 @@ module Fog
           Fog::Logger.debug "Copying blob: x-ms-copy-id: #{copy_id}, x-ms-copy-status: #{copy_status}"
           [copy_id, copy_status]
         end
+
+        def copy_object(source_container, source_blob, target_container, target_blob, options = {})
+          copy_blob(target_container, target_blob, source_container, source_blob, options)
+        end
       end
 
       # This class provides the mock implementation for unit tests.
@@ -24,6 +28,10 @@ module Fog
         def copy_blob(*)
           %w(abc123 pending)
         end
+
+        def copy_object(*)
+          %w(abc123 pending)
+        end
       end
     end
   end
```

**Comment by mshibuya** [comment]
It's fixed now, since #2583 changed to use File#copy which is also implemented in Fog::Storage::AzureRM::File.
https://github.com/fog/fog-azure-rm/blob/9da5cc04189cd1ec751c5bc1e2f5406ce6b179d0/lib/fog/azurerm/models/storage/file.rb#L152
